flac.md

```diff
@@ -630,6 +630,28 @@ Decoding the coded residual thus involves selecting the right coding method, fin
 
 Following the last subframe is the frame footer. If the last subframe is not byte aligned (i.e. the bits required to store all subframes put together are not divisible by 8), zero bits are added until byte alignment is reached. Following this is a 16-bit CRC, initialized with 0, with polynomial x^16 + x^15 + x^2 + x^0. This CRC covers the whole frame excluding the 16-bit CRC, including the sync code.
 
+# Container mappings
+
+The FLAC format can be used without any container as the FLAC format already provides for a very thin transport layer. However, the functionalty of this transport is rather limited, and to be able to combine FLAC audio with video, it needs to be encapsulated by a more capable container. This presents a problem: the transport layer provided by the FLAC format mixes data that belongs to the encoded data (like blocksize and sample rate) with data that belongs to the transport (like checksum and timecode). The choice was made to encapsulate FLAC frames as they are, which means some data will be duplicated and potentially deviating between the FLAC frames and the encapsulating container.
+
+To differentiate encapsulated FLAC streams from non-encapsulated FLAC streams, the latter will be referred to as 'native FLAC'.
+
+## Ogg mapping
+
+The Ogg container format is defined in [@?RFC3533]. The first packet of a logical bitstream carrying FLAC data contains the bytes `0x7F 0x46 0x4C 0x41 0x43` as defined by [@?RFC5334], followed by two bytes coding the version number of the FLAC-in-Ogg mapping. These bytes are `0x01 0x00`, meaning version 1.0 of the mapping. Following the version number bytes is a 2-byte unsigned number coded big-endian containing the number of header packets (i.e. the number of metadata blocks) that follow the first packet.  This number MAY be 0, which means the number of packets that follow is unknown. Following the 2-byte number of header packets is the `fLaC` signature and a streaminfo metadata block. Taken together this means the first packet of an logical stream of FLAC-in-Ogg is always 79 bytes. This first packet MUST NOT share a Ogg page with any other packets.
+
+Following the first packet are one or more header packets i.e. metadata blocks. The first of these MUST be a vorbis comment metadata block. This is contrary to native FLAC streams, where the order of metadata blocks is not important except for the streaminfo block and where a vorbis comment metadata block is optional.
```

Comment:
Why is a vorbis comment metadata block mandatory and why the order is important?

Reply:
That is part of the definition, so I don't know why

Reply:
Wondering if any decoder suffers if it is out of order and/or without vorbis comment block. It may be better to avoid such constraints if not useful, IMO.

Reply:
I need to do some digging to find out whether there is any specific reason for this

Reply:
It seems the reasoning for this is here: https://lists.xiph.org/pipermail/flac-dev/2004-July/001567.html

In retrospect this reasoning seems a little weak: it is so the vendor string has a fixed offset from the start of the Ogg file and a decoder can use this to figure out whether it can decode a particular FLAC file. It seems the vendor string was being used as a version string instead, which it isn't of course.

> what is your recommendation here?  is it that the fLaC and
first STREAMINFO (which holds sample rate, #samples,
resolution, etc) packets be flushed together so that they
are in the same page, page 0?  the total size of both packets
is 42 bytes so this seems to be no problem.
> 
> but adding in the codec version to the first page.... in
FLAC the codec version is in the vendor tag of the vorbis
comments.  the vorbis comments are in their own metadata
block/packet, but there is no requirement for the vorbis
comments to follow the STREAMINFO immediately.  there may
be other metadata in between.  now, I can enforce it in
libOggFLAC that it follows STREAMINFO immediately, but then
the question is, how can you guarantee that the whole vorbis
comment packet will also fit in page 0, given that there is
not much restriction on the size of comments?  is it enough
that the first part of the vorbis comment packet that contains
the vendor string is in page 0?

This might have worked back in 2004, but we have many more FLAC encoders these days, and they don't follow the same versioning as libFLAC.

So, that seems to be the why. As for whether this results in problems if omitted, that is a much harder question to answer.

I agree that useless constraints should be removed, but I don't know what we're risking with this. It doesn't really seem to hurt, it is simply a minor inconvenience to muxers. I see no reason not to leave it in.

FFMPEG does it too
https://github.com/FFmpeg/FFmpeg/blob/ed83a3a5bddf4c209157dc9f041eda0721b4c3e0/libavformat/oggenc.c#L328-L329

Reply:
> I see no reason not to leave it in.

It seems weird to me that a VORBIS comment is mandatory, i.e. it is legitimate not to have one (it is optional in FLAC native stream, why not in OGG?). Maybe a SHOULD there, but not blocking for me, just annoyed by "useless" constraints.

Reply:
Then perhaps something like:
```

Reply:
Fine for me.